### PR TITLE
Deterministic InnerClass attribute under joint/separate compilation

### PR DESCRIPTION
**Pull request by retronym**
Use the private flag as defined in source code, rather than the potentially
less-private status of the class at the end of compiler pipeline.

Fixes scala/bug#12085

### CODE REVIEW

**Comment by retronym** [comment]
Checkfile updated, the new behaviour emits sorted declared inner classes first, followed by sorted referred inner classes.

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala [hunk 1]
```diff
@@ -854,7 +856,7 @@ abstract class BTypes {
    */
   final case class ClassInfo(superClass: Option[ClassBType], interfaces: List[ClassBType], flags: Int,
                              nestedClasses: Lazy[List[ClassBType]], nestedInfo: Lazy[Option[NestedInfo]],
-                             inlineInfo: InlineInfo)
+                             inlineInfo: InlineInfo, exitingTyperPrivate: Boolean)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala]: This field would be better suited in the `nestedInfo` class, which already has an `isStaticNestedClass: Boolean`

File: test/files/jvm/t8582.check [hunk 2]
```diff
@@ -25,8 +25,8 @@ The module members are not in the InnerClass table of the module class (unless r
 An outer class has a InnerClass attribute for direct nested classes
  className[A1$B1] outerClassName[A1] innerName[B1] access[1]
 A nested class has an InnerClass attribute for itself (and also for its nested classes)
- className[A1$B1] outerClassName[A1] innerName[B1] access[1]
  className[A1$B1$C1] outerClassName[A1$B1] innerName[C1] access[1]
+ className[A1$B1] outerClassName[A1] innerName[B1] access[1]
```
Comment (lrytz) [inline test/files/jvm/t8582.check]: There's a comment somewhere saying
```
// sorting ensures nested classes are listed after their enclosing class thus satisfying the Eclipse Java compiler
```
This change seems to break that property, no?

**Review by lrytz — CHANGES_REQUESTED** [review]
I'm not sure we should remain bug compatible here. It's clearly a bug, I think we can fix it in a minor release. In the current setup, I'm a little worried about the fact that the inner class entry for a class is different depending on where it appears.

File: test/files/jvm/t8582.check (hunk 2 above)
Comment (retronym) [reply test/files/jvm/t8582.check]: Declared classes are still sorted, as a referenced classes which would still satisfy that requirement. But to be conservative I'll restore the global sorting.

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala (hunk 1 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala]: Makes sense, done.

**Review by lrytz — APPROVED** [review]

**Pull request merged**